Lib/tty.py

```diff
@@ -15,22 +23,136 @@
+    # or a signal is received.
     mode[CC][VMIN] = 1
     mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
 
-def setcbreak(fd, when=TCSAFLUSH):
-    """Put terminal into a cbreak mode."""
-    mode = tcgetattr(fd)
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON)
+def cfmakecbreak(mode):
+    """cbreak mode termios"""
+    # Do not map CR to NL on input.
+    mode[IFLAG] &= ~(ICRNL)
+
+    # Do not echo characters; disable canonical input.
+    mode[LFLAG] &= ~(ECHO | ICANON)
+
+    # POSIX.1-2017, 11.1.7 Non-Canonical Mode Input Processing,
+    # Case B: MIN>0, TIME=0
+    # A pending read shall block until MIN (here 1) bytes are received,
+    # or a signal is received.
     mode[CC][VMIN] = 1
     mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
+
+def setraw(fd, when=TCSAFLUSH):
+    """Put terminal into raw mode.
+    Returns original termios."""
+    mode = tcgetattr(fd)
+    new = list(mode)
+    cfmakeraw(new)
+    tcsetattr(fd, when, new)
+    return mode
+
+def setcbreak(fd, when=TCSAFLUSH):
+    """Put terminal into cbreak mode.
+    Returns original termios."""
+    mode = tcgetattr(fd)
+    new = list(mode)
+    cfmakecbreak(new)
+    tcsetattr(fd, when, new)
+    return mode
+
+try:
+    from termios import TIOCGWINSZ, TIOCSWINSZ
+    HAVE_WINSZ = True
+
+    class winsize:
+
+        def __init__(self, row=0, col=0, xpixel=0, ypixel=0, fd=None):
+            """Creates winsize object.
+            If fd is not None, gets window size of tty of which fd is a
+            file descriptor. If fd is None, creates the winsize object
+            based on the row, col, xpixel, and ypixel values provided."""
+            if fd != None:
+                self.tcgetwinsize(fd)
+            else:
+                self.ws_row = row
+                self.ws_col = col
+                self.ws_xpixel = xpixel
+                self.ws_ypixel = ypixel
+
+        def __setattr__(self, name, value):
+            if not isinstance(value, int) or value < 0:
+                raise TypeError("expected nonnegative integer")
+            else:
+                super().__setattr__(name, value)
+
+        def __eq__(self, obj):
+            if isinstance(obj, self.__class__):
+                return self.__dict__ == obj.__dict__
+            else:
+                return False
+
+        def __repr__(self):
+            return f"winsize(ws_row={self.ws_row}, ws_col={self.ws_col}, ws_xpixel={self.ws_xpixel}, ws_ypixel={self.ws_ypixel})"
+
+        def __str__(self):
+            return f"(row={self.ws_row}, col={self.ws_col}, xpixel={self.ws_xpixel}, ypixel={self.ws_ypixel})"
+
+        def tcgetwinsize(self, fd):
+            """Gets window size of tty of which fd is a file descriptor."""
+            s = pack("HHHH", 0, 0, 0, 0)
+            try:
+                w = unpack("HHHH", ioctl(fd, TIOCGWINSZ, s))
+            except OSError as e:
+                raise error(e.errno, os.strerror(e.errno)) from e # termios.error
```

Comment:
The original author intended this to be an extension of the `termios` module. In fact, notice that there is a `from termios import *` at the beginning of the file; the author also uses `tty.error` instead of `termios.error` in "Lib/pty.py". Therefore, all functions/methods must raise `termios.error` [ which is the same as `tty.error` ] when an `ioctl()`/`tcgetattr()`/`tcsetattr()`/... on `fd` [ which is expected to be the fd of a tty ] fails. Anyway, as mentioned before, `tcgetwinsize()` and `tcsetwinsize()` declarations are set to appear in `<termios.h>` as a part of IEEE Std 1003.1 ("POSIX.1") issue 8 [ the upcoming version of POSIX ]; see https://www.austingroupbugs.net/view.php?id=1151#c3856. Therefore, when our `winsize.tcgetwinsize()` and `winsize.tcsetwinsize()` methods fail, `termios.error` must be raised. That was one of the reasons why I removed `login_tty()` from this file; it has got almost nothing to do with `termios`. I will include it in the next pull request as a part of a different file.